server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java

```diff
@@ -177,16 +215,27 @@ public synchronized FetchResult<T> fetchData(DiscoveryNodes nodes, Set<String> i
                     }
                 }
             }
-            Set<String> allIgnoreNodes = unmodifiableSet(new HashSet<>(nodesToIgnore));
+
+            Map<ShardId, Set<String>> allIgnoreNodesMap = unmodifiableMap(new HashMap<>(shardToIgnoreNodes));
             // clear the nodes to ignore, we had a successful run in fetching everything we can
             // we need to try them if another full run is needed
-            nodesToIgnore.clear();
+            shardToIgnoreNodes.clear();
             // if at least one node failed, make sure to have a protective reroute
             // here, just case this round won't find anything, and we need to retry fetching data
-            if (failedNodes.isEmpty() == false || allIgnoreNodes.isEmpty() == false) {
-                reroute(shardId, "nodes failed [" + failedNodes.size() + "], ignored [" + allIgnoreNodes.size() + "]");
+            // If ignore node even for a single shard in batch of shards then also do a reroute
+            if (failedNodes.isEmpty() == false
+                || allIgnoreNodesMap.values().stream().anyMatch(ignoreNodeSet -> ignoreNodeSet.isEmpty() == false)) {
```

Comment:
nit: Remove explicit `false` check and use `entrySet` in place of `values`
```
allIgnoreNodesMap.entrySet()
    .stream()
    .anyMatch(entry -> !entry.getValue().isEmpty());
```

Reply:
nit: Just trying to maintain to maintain same coding pattern thats used in class, for readability [I think] the authors of this class[and other classes in the package] have used negation statements with `== false` at most of the places.

Reply:
I understand this is more of personal preference at this point. I find `if (!x)` more readable than ` if (x == false)`.